stdlib/join/merge_join.go

```diff
@@ -0,0 +1,518 @@
+		return position, true
+	}
+
+	// The only condition where we would not emit a signal to join and flush
+	// the list of products is if we received a bunch of `joinRows` all
+	// from the same side.
+	//
+	// So if the previous product is only populated on one side, then it is either the first
+	// item in the list, or all of the items before it are also populated on the same side.
+	canJoin := false
+	if position > 0 {
+		position--
+		prev := s.products[position]
+		if isLeft {
+			canJoin = prev.left.len() == 0 && prev.right.len() > 0
+		} else {
+			canJoin = prev.left.len() > 0 && prev.right.len() == 0
+		}
+	}
+	return position, canJoin
+}
+
+func (s *joinState) join(method string, fn *JoinFn, joinable int) ([]table.Chunk, error) {
+	if fn.compiled == nil {
+		err := fn.Prepare(s.left.schema, s.right.schema)
+		if err != nil {
+			return nil, err
+		}
+	}
+	joined := make([]table.Chunk, 0, joinable+1)
+	for i := 0; i <= joinable; i++ {
+		p, _ := s.products[i].evaluate(method, *fn)
+		joined = append(joined, *p)
+	}
+	return joined, nil
+}
+
+func (s *joinState) finished() bool {
+	return s.left.done && s.right.done
+}
+
+type sideState struct {
+	schema   []flux.ColMeta
+	keyCols  []flux.ColMeta
+	chunks   []table.Chunk
+	keyEnd   int
+	keyStart int // index of the first row in chunks[0] that contained the current join key
+	done     bool
+}
+
+func (s *sideState) scan(c table.Chunk) (*joinKey, joinRows) {
+	if len(s.schema) == 0 {
+		s.schema = c.Cols()
+	}
+	if len(s.keyCols) == 0 {
+		s.keyCols = c.Key().Cols()
+	}
+	key, complete := s.advance(c)
+	if !complete {
+		s.addChunk(c)
+		return nil, nil
+	}
+	return key, s.consumeRows(c)
+}
+
+func (s *sideState) addChunk(c table.Chunk) {
+	s.chunks = append(s.chunks, c)
+}
+
+// advance iterates over each row of c until it either finds a new join key or
+// reaches the end of the chunk. Returns true if it finds the end of a join key,
+// along with the key itself
+func (s *sideState) advance(c table.Chunk) (*joinKey, bool) {
+	startKey := joinKeyFromRow(s.keyCols, c, s.keyEnd)
+	key := startKey
+
+	for ; s.keyEnd < c.Len(); s.keyEnd++ {
+		key = joinKeyFromRow(s.keyCols, c, s.keyEnd)
+		if !key.equal(startKey) {
+			s.keyStart = s.keyEnd
+			return &startKey, true
+		}
+	}
+	return nil, false
+}
+
+// collects rows up to the indices set by `scan()` and returns them in one
+// data structure. It will discard any exhausted chunks.
+func (s *sideState) consumeRows(c table.Chunk) joinRows {
+	if len(s.chunks) == 0 {
+		return joinRows([]table.Chunk{getChunkSlice(c, s.keyStart, s.keyEnd)})
+	}
+	s.chunks[0] = getChunkSlice(c, s.keyStart, c.Len())
+	s.chunks = append(s.chunks, getChunkSlice(c, 0, s.keyEnd))
+
+	// TODO(sean): I think rows may just point to the underlying data of s.chunks, rather than
+	// a copy of it. If so, setting s.chunks to an empty slice will cause problems. Will confirm
+	// this during testing.
+	rows := s.chunks
+	s.chunks = s.chunks[:0]
+	return rows
+}
```

Comment:
Just now noticing some major flaws in these two functions. Working on a fix now.